### ISSUE TITLE
Any optimization at the neg_log_likelihood computation?

### ISSUE THREAD

**Issue by Wu-Jingcheng** [chqiwang/convseg#10]
Thanks for your brilliant job.

I am currently looking at building it on my own with pytorch.

Sorry that  I am relatively new to NLP and thereby (maybe) immature on some issue. When the loss function looks like computationally expensive. Like, if I have a 28-char sentence and apply 4 tags as yours in the paper. I would definitely have 4^28 = 72057594037927936 results to take in. Is it really so expensive? Or there might be some form of optimization to avoid the daunting overhead?

**Comment by Wu-Jingcheng** [comment]
Never mind. 

I've looked at some documents and implementation examples, some of which are listed as fellows:
https://towardsdatascience.com/implementing-a-linear-chain-conditional-random-field-crf-in-pytorch-16b0b9c4b4ea
https://pytorch.org/tutorials/beginner/nlp/advanced_tutorial.html#sphx-glr-beginner-nlp-advanced-tutorial-py

I would close it.

Thank you again.

**Comment by lvjiujin** [comment]
> Thanks for your brilliant job.
> 
> I am currently looking at building it on my own with pytorch.
> 
> Sorry that I am relatively new to NLP and thereby (maybe) immature on some issue. When the loss function looks like computationally expensive. Like, if I have a 28-char sentence and apply 4 tags as yours in the paper. I would definitely have 4^28 = 72057594037927936 results to take in. Is it really so expensive? Or there might be some form of optimization to avoid the daunting overhead?

can you implement this paper with pytorch? I also study the paper, but I want to see the pytorch code.
